File: ios/Wayfare/AddExpenseSheet.swift
import SwiftUI
import WayfareCore

/// Applies one keypad press to the amount. Kept pure and separate so the rules —
/// one decimal point, at most two decimals, no runaway leading zeroes — are
/// readable in one place rather than scattered through the key handlers.
/// Mirrors `applyKey` in the Android `AddExpenseSheet.kt`.
func applyKey(_ current: String, _ key: String) -> String {
  if key == "⌫" { return String(current.dropLast()) }
  if key == "." {
    if current.contains(".") { return current }
    return current.isEmpty ? "0." : current + "."
  }
  if current.contains("."), current.split(separator: ".", maxSplits: 1).last?.count ?? 0 >= 2 {
    return current
  }
  if current == "0" { return key }
  if current.count >= 12 { return current }
  return current + key
}

/// Adding a line. The amount is typed on a keypad rather than a keyboard,
/// because an amount is the one field that is always a number and always the
/// first thing you know — the sheet opens straight onto it.
///
/// The design's payer control ("You paid / Sam paid / Split it") is deliberately
/// absent: this is one person's tape, and there is nobody to split with.
struct AddExpenseSheet: View {
  @Environment(AppStore.self) private var store
  @Environment(\.dismiss) private var dismiss
  let trip: Trip
  let expense: Expense?

  @State private var amount = ""
  @State private var title = ""
  @State private var paidIn = ""
  @State private var category: WayfareCore.Category = .food
  @State private var spentOn = Day.today
  @State private var note = ""
  @State private var showNote = false
  @State private var pickingDate = false
  @State private var busy = false
  @State private var error: String?
  @State private var confirmDelete = false
  @State private var loaded = false

  /// Shared with the trip form so the keypad accepts exactly what the ledger does.
  private var parsed: Decimal? { parseAmount(amount).flatMap { $0 > 0 ? $0 : nil } }
  private var rate: Decimal? {
    // Changing a note, category or date must not revalue a historical cost, so
    // an edit that leaves the currency alone keeps the rate saved with the line.
    if let expense, paidIn == (expense.originalCurrency ?? trip.currency) {
      return expense.fxRate ?? (paidIn == trip.currency ? 1 : nil)
    }
    return rateBetween(from: paidIn, to: trip.currency, rates: store.fx.rates)
  }
  private var converted: Decimal? {
    guard let parsed, let rate else { return nil }
    return rounded(parsed * rate)
  }
  private var currencies: [String] {
    Array(Set(store.fx.rates.keys).union([paidIn, trip.currency])).sorted()
  }

  var body: some View {
    VStack(spacing: 0) {
      HStack {
        Text(expense == nil ? "Add a line" : "Fix the line").typeStyle(.headlineSmall)
        Spacer()
        CircleIconButton(symbol: "xmark", label: "Close", raised: true) { dismiss() }
      }.padding(.top, 12)

      if let error { Notice(text: error, isError: true).padding(.top, 14) }

      // The figure, set the way a till prints it: big, amber, monospaced.
      VStack(spacing: 10) {
        Text(symbolFor(paidIn) + (amount.isEmpty ? "0" : amount))
          .font(.custom(Face.monoMedium, size: 54, relativeTo: .largeTitle)).tracking(-2)
          .foregroundStyle(amount.isEmpty ? Palette.amber.opacity(0.45) : Palette.amber)
          .lineLimit(1).minimumScaleFactor(0.5)
          .accessibilityLabel("Amount \(amount.isEmpty ? "zero" : amount) \(paidIn)")
        // The title sits where a receipt puts the merchant: under the figure.
        TextField(
          "", text: $title,
          prompt: Text("What was it for?").foregroundStyle(Palette.slate)
        )
        .typeStyle(.titleMedium).foregroundStyle(Palette.paper)
        .multilineTextAlignment(.center)
        if paidIn != trip.currency, let converted {
          Text(
            "≈ \(money(converted, trip.currency)) · 1 \(paidIn) = \(rate.map { "\(rounded($0, scale: 4))" } ?? "—") \(trip.currency)"
              + (store.fx.stale ? " · offline rate" : "")
          )
          .typeStyle(.bodySmall).foregroundStyle(Palette.slate).multilineTextAlignment(.center)
        }
      }.padding(.top, 18)

      // Category first, because it is the one chip you always set.
      ScrollView(.horizontal, showsIndicators: false) {
        HStack(spacing: 8) {
          ForEach(WayfareCore.Category.allCases, id: \.self) { item in
            FilterPill(label: item.label, selected: category == item) { category = item }
          }
        }
      }.padding(.top, 14)

      // Then the three facts the design leaves out but the ledger needs.
      HStack(spacing: 8) {
        FilterPill(label: dateLabel, selected: spentOn != Day.today) { pickingDate = true }
        Menu {
          Picker("", selection: $paidIn) {
            ForEach(currencies, id: \.self) { Text($0).tag($0) }
          }
        } label: {
          Text(paidIn).typeStyle(.bodySmall).foregroundStyle(Palette.steel)
            .frame(maxWidth: .infinity).frame(height: 34)
            .overlay(
              RoundedRectangle(cornerRadius: Radius.chip).stroke(Palette.outline, lineWidth: 1))
        }
        FilterPill(label: note.isEmpty ? "Note" : "Note ✓", selected: showNote) {
          showNote.toggle()
        }
      }.padding(.top, 10)

      if showNote {
        TextField(
          "", text: $note,
          prompt: Text("Anything worth remembering").foregroundStyle(Palette.slate),
          axis: .vertical
        )
        .typeStyle(.bodyMedium).foregroundStyle(Palette.paper).lineLimit(1...3)
        .padding(.horizontal, 14).padding(.vertical, 12)
        .overlay(RoundedRectangle(cornerRadius: Radius.control).stroke(Palette.outline, lineWidth: 1))
        .padding(.top, 10)
      }

      keypad.padding(.top, 14)

      PrimaryButton(
        title: expense == nil ? "Print it to the tape" : "Save the line", busy: busy, action: save
      ).padding(.top, 14)

      if expense != nil {
        Button("Delete this line", role: .destructive) { confirmDelete = true }
          .typeStyle(.bodyMedium).fontWeight(.semibold).foregroundStyle(Palette.errorRed)
          .padding(.top, 10)
      }
      Spacer(minLength: 12)
    }
    .padding(.horizontal, 20)
    .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .top)
    .background(Palette.card)
    .presentationDetents([.large])
    .presentationCornerRadius(Radius.sheet)
    .presentationBackground(Palette.card)
    .interactiveDismissDisabled(busy)
    .sheet(isPresented: $pickingDate) { datePicker }
    .confirmationDialog(
      "Delete this line?", isPresented: $confirmDelete, titleVisibility: .visible
    ) {
      Button("Delete line", role: .destructive) {
        guard let expense else { return }
        run {
          try await store.deleteExpense(expense)
          dismiss()
        }
      }
    } message: {
      Text("This cannot be undone.")
    }
    .onAppear(perform: load)
  }

  private var dateLabel: String {
    switch spentOn {
    case Day.today: "Today"
    case Day.string(Date().addingTimeInterval(-86400)): "Yesterday"
    default: shortDate(spentOn) ?? spentOn
    }
  }

  private var keypad: some View {
    VStack(spacing: 10) {
      ForEach(Array(keys.chunked(into: 3).enumerated()), id: \.offset) { _, row in
        HStack(spacing: 10) {
          ForEach(row, id: \.self) { key in
            Button { amount = applyKey(amount, key) } label: {
              Text(key).font(.custom(Face.mono, size: 22, relativeTo: .title2))
                .foregroundStyle(Palette.paper)
                .frame(maxWidth: .infinity).frame(height: 52)
                .background(Palette.cardRaised, in: RoundedRectangle(cornerRadius: Radius.key))
            }.buttonStyle(.plain)
          }
        }
      }
    }
  }

  private var keys: [String] { ["1", "2", "3", "4", "5", "6", "7", "8", "9", ".", "0", "⌫"] }

  private var datePicker: some View {
    NavigationStack {
      DatePicker(
        "Date",
        selection: Binding(
          get: { Day.date(spentOn) ?? Date() },
          set: { spentOn = Day.string($0) }),
        displayedComponents: .date
      )
      .datePickerStyle(.graphical).tint(Palette.amber).padding()
      .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .top)
      .canvasScreen().navigationTitle("When was it?").navigationBarTitleDisplayMode(.inline)
      .toolbar {
        ToolbarItem(placement: .confirmationAction) {
          Button("Done") { pickingDate = false }
        }
      }
    }.presentationDetents([.medium])
  }

  private func load() {
    guard !loaded else { return }
    loaded = true
    let remembered = UserDefaults.standard.string(forKey: Self.paidInKey(trip: trip, store: store))
    paidIn = expense?.originalCurrency ?? (expense == nil ? remembered : nil) ?? trip.currency
    if let expense {
      let shown = expense.originalAmount ?? expense.amount
      amount = "\(shown)"
      title = expense.title
      category = expense.category
      spentOn = expense.spentOn
      note = expense.note ?? ""
      showNote = !(expense.note ?? "").isEmpty
    }
  }

  private static func paidInKey(trip: Trip, store: AppStore) -> String {
    "paidIn.\(store.userId ?? "").\(trip.id)"
  }

  private func save() {
    error =
      if parsed == nil {
        "Enter an amount above zero."
      } else if title.trimmingCharacters(in: .whitespaces).isEmpty {
        "What was it for?"
      } else if paidIn != trip.currency && rate == nil {
        "No rate is available for \(paidIn) to \(trip.currency)."
      } else {
        nil
      }
    guard error == nil, let parsed else { return }

    var draft = expense ?? Expense(tripId: trip.id, userId: store.userId ?? "")
    draft.title = title.trimmingCharacters(in: .whitespaces)
    draft.amount = rounded(converted ?? parsed)
    draft.originalAmount = paidIn == trip.currency ? nil : rounded(parsed)
    draft.originalCurrency = paidIn == trip.currency ? nil : paidIn
    draft.fxRate = paidIn == trip.currency ? nil : rate
    draft.category = category
    draft.spentOn = spentOn
    draft.note = note.trimmingCharacters(in: .whitespaces).isEmpty ? nil : note
    draft.syncState = .synced
    if expense == nil { draft.createdAt = ISO8601DateFormatter().string(from: Date()) }

    let isNew = expense == nil
    run {
      try await store.saveExpense(draft, isNew: isNew)
      UserDefaults.standard.set(paidIn, forKey: Self.paidInKey(trip: trip, store: store))
      dismiss()
    }
  }

  private func run(_ action: @escaping () async throws -> Void) {
    busy = true
    Task {
      defer { busy = false }
      do { try await action() } catch { self.error = error.localizedDescription }
    }
  }
}

/// A line someone else added: readable, never editable.
struct ReadOnlyExpenseSheet: View {
  @Environment(AppStore.self) private var store
  @Environment(\.dismiss) private var dismiss
  let expense: Expense
  let currency: String
  var body: some View {
    NavigationStack {
      ScrollView {
        VStack(alignment: .leading, spacing: 12) {
          FigureWithQualifier(figure: money(expense.amount, currency), size: 38)
          Text(
            "Printed by \(store.name(for: expense.userId)). Only the traveller who added a line can change it."
          ).typeStyle(.bodyMedium).foregroundStyle(Palette.slate)
          MonoLabel(
            text: "\(expense.category.label.uppercased()) · \(tapeDayLabel(expense.spentOn))",
            color: Palette.slate)
          if let original = expense.originalCurrency, let amount = expense.originalAmount {
            Text("Originally \(amount) \(original)").typeStyle(.bodyMedium)
              .foregroundStyle(Palette.slate)
          }
          if let note = expense.note { Text(note).typeStyle(.bodyMedium) }
        }.frame(maxWidth: .infinity, alignment: .leading).padding(20)
      }
      .canvasScreen().navigationTitle(expense.title).navigationBarTitleDisplayMode(.inline)
      .toolbar { ToolbarItem(placement: .confirmationAction) { Button("Close") { dismiss() } } }
    }.presentationDetents([.medium])
  }
}

/// A line that never reached the server: kept, explained, retried or discarded.
struct UnsyncedExpenseSheet: View {
  @Environment(AppStore.self) private var store
  @Environment(\.dismiss) private var dismiss
  let expense: Expense
  let currency: String
  @State private var busy = false
  @State private var error: String?
  var body: some View {
    let failed = expense.syncState == .failed
    NavigationStack {
      ScrollView {
        VStack(alignment: .leading, spacing: 14) {
          DottedLeaderRow(label: expense.title, amount: money(expense.amount, currency))
          Text(
            failed
              ? "It is kept here so nothing is lost, but it is not counted in the trip total and other travellers cannot see it yet."
              : "It is saved on this phone and will sync as soon as there is a connection. You can edit it once it lands."
          ).typeStyle(.bodyMedium).foregroundStyle(Palette.slate)
          if let reason = expense.syncError { Notice(text: reason, isError: failed) }
          if let error { Notice(text: error, isError: true) }
          PrimaryButton(title: "Try again", busy: busy) {
            run {
              try await store.retryExpense(expense)
              dismiss()
            }
          }
          if failed {
            Button("Discard this line", role: .destructive) {
              run {
                try await store.discardExpense(expense)
                dismiss()
              }
            }
            .typeStyle(.bodyMedium).fontWeight(.semibold).foregroundStyle(Palette.errorRed)
            .frame(maxWidth: .infinity)
          }
        }.padding(20)
      }
      .canvasScreen().disabled(busy)
      .navigationTitle(failed ? "This line did not print" : "Still printing")
      .navigationBarTitleDisplayMode(.inline)
      .toolbar { ToolbarItem(placement: .cancellationAction) { Button("Close") { dismiss() } } }
    }.presentationDetents([.medium])
  }
  private func run(_ action: @escaping () async throws -> Void) {
    busy = true
    Task {
      defer { busy = false }
      do { try await action() } catch { self.error = error.localizedDescription }
    }
  }
}

extension Array {
  func chunked(into size: Int) -> [[Element]] {
    stride(from: 0, to: count, by: size).map { Array(self[$0..<Swift.min($0 + size, count)]) }
  }
}

File: android/app/src/main/java/com/wayfare/app/ui/AddExpenseSheet.kt
package com.wayfare.app.ui

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Close
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.SheetValue
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.rememberUpdatedState
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.wayfare.app.core.Category
import com.wayfare.app.core.Expense
import com.wayfare.app.core.ExpenseDraft
import com.wayfare.app.core.money
import com.wayfare.app.core.symbolFor
import com.wayfare.app.data.FxRepository
import kotlinx.coroutines.launch
import java.math.RoundingMode
import java.time.LocalDate

/**
 * Adding a line. The amount is typed on a keypad rather than a keyboard, because
 * an amount is the one field that is always a number and always the first thing
 * you know — the sheet opens straight onto it.
 *
 * The design's payer control ("You paid / Sam paid / Split it") is deliberately
 * absent: this is one person's tape, and there is nobody to split with.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun AddExpenseSheet(
    expense: Expense? = null,
    currency: String,
    initialPaidIn: String = currency,
    fx: FxRepository,
    onDismiss: () -> Unit,
    onSave: suspend (ExpenseDraft) -> Result<*>,
    onDelete: (suspend () -> Result<*>)? = null,
    onRememberCurrency: suspend (String) -> Unit = {},
) {
    val snapshot by fx.snapshot.collectAsStateWithLifecycle()
    var amount by rememberSaveable(expense?.id) {
        mutableStateOf((expense?.originalAmount ?: expense?.amount)?.stripTrailingZeros()?.toPlainString().orEmpty())
    }
    var title by rememberSaveable(expense?.id) { mutableStateOf(expense?.title.orEmpty()) }
    var paidIn by rememberSaveable(expense?.id, initialPaidIn) { mutableStateOf(expense?.originalCurrency ?: initialPaidIn) }
    var category by rememberSaveable(expense?.id) { mutableStateOf(expense?.category ?: Category.Food) }
    var spentOn by rememberSaveable(expense?.id) { mutableStateOf(expense?.spentOn ?: LocalDate.now()) }
    var note by rememberSaveable(expense?.id) { mutableStateOf(expense?.note.orEmpty()) }
    var showNote by rememberSaveable(expense?.id) { mutableStateOf(!expense?.note.isNullOrBlank()) }
    var error by rememberSaveable { mutableStateOf<String?>(null) }
    var busy by remember { mutableStateOf(false) }
    var confirmDelete by remember { mutableStateOf(false) }
    var discard by remember { mutableStateOf(false) }
    val scope = rememberCoroutineScope()

    val parsed = amount.toBigDecimalOrNull()
    val converted = parsed?.let { fx.convert(it, paidIn, currency, snapshot) }
    val dirty = title != expense?.title.orEmpty() ||
        amount != (expense?.originalAmount ?: expense?.amount)?.stripTrailingZeros()?.toPlainString().orEmpty() ||
        paidIn != (expense?.originalCurrency ?: initialPaidIn) || category != (expense?.category ?: Category.Food) ||
        spentOn != (expense?.spentOn ?: LocalDate.now()) || note != expense?.note.orEmpty()

    fun dismiss() {
        if (busy) return
        if (dirty) discard = true else onDismiss()
    }

    val currentCanDismiss by rememberUpdatedState(!dirty && !busy)
    val sheetState = rememberModalBottomSheetState(
        skipPartiallyExpanded = true,
        confirmValueChange = { value ->
            if (value == SheetValue.Hidden && !currentCanDismiss) {
                discard = true
                false
            } else true
        },
    )

    ModalBottomSheet(
        onDismissRequest = ::dismiss,
        sheetState = sheetState,
        sheetMaxWidth = Dp.Unspecified,
        containerColor = CardNavy,
        contentColor = Paper,
        shape = RoundedCornerShape(topStart = 32.dp, topEnd = 32.dp),
        scrimColor = NightDeep.copy(alpha = .72f),
    ) {
        Column(Modifier.fillMaxWidth().padding(horizontal = 20.dp).navigationBarsPadding().imePadding()) {
            Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                Text(
                    if (expense == null) "Add a line" else "Fix the line",
                    Modifier.weight(1f),
                    style = MaterialTheme.typography.headlineSmall,
                )
                CircleIconButton(Icons.Outlined.Close, "Close", raised = true, onClick = ::dismiss)
            }

            error?.let { Notice(it, true, Modifier.padding(top = 14.dp)) }

            // The figure, set the way a till prints it: big, amber, monospaced.
            Column(
                Modifier.fillMaxWidth().padding(top = 18.dp, bottom = 4.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                Text(
                    symbolFor(paidIn).trim() + amount.ifBlank { "0" },
                    color = if (amount.isBlank()) Amber.copy(alpha = .45f) else Amber,
                    fontFamily = PlexMono,
                    fontWeight = FontWeight.Medium,
                    fontSize = 54.sp,
                    lineHeight = 56.sp,
                    letterSpacing = (-2).sp,
                    maxLines = 1,
                )
                // The title sits where a receipt puts the merchant: under the figure.
                BasicTextField(
                    value = title,
                    onValueChange = { title = it.take(120) },
                    modifier = Modifier.fillMaxWidth().padding(top = 10.dp),
                    textStyle = MaterialTheme.typography.titleMedium.copy(
                        color = Paper,
                        textAlign = TextAlign.Center,
                    ),
                    singleLine = true,
                    cursorBrush = SolidColor(Amber),
                    decorationBox = { field ->
                        Box(contentAlignment = Alignment.Center) {
                            if (title.isEmpty()) {
                                Text(
                                    "What was it for?",
                                    color = Slate,
                                    style = MaterialTheme.typography.titleMedium,
                                )
                            }
                            field()
                        }
                    },
                )
                if (paidIn != currency && converted != null) {
                    Text(
                        "≈ ${money(converted, currency)} · 1 $paidIn = ${fx.rateBetween(paidIn, currency, snapshot)?.setScale(4, RoundingMode.HALF_UP)} $currency" +
                            if (snapshot.stale) " · offline rate" else "",
                        Modifier.padding(top = 6.dp),
                        color = Slate,
                        style = MaterialTheme.typography.bodySmall,
                        textAlign = TextAlign.Center,
                    )
                }
            }

            // Category first, because it is the one chip you always set.
            Row(
                Modifier.fillMaxWidth().horizontalScroll(rememberScrollState()).padding(top = 14.dp),
                horizontalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                Category.entries.forEach { item ->
                    FilterPill(item.label, category == item) { category = item }
                }
            }

            // Then the three facts the design leaves out but the ledger needs.
            Row(
                Modifier.fillMaxWidth().padding(top = 10.dp),
                horizontalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                DatePill(spentOn, Modifier.weight(1f)) { spentOn = it }
                CurrencyPill(paidIn, Modifier.weight(1f)) { paidIn = it }
                FilterPill(
                    if (note.isBlank()) "Note" else "Note ✓",
                    showNote,
                    Modifier.weight(1f),
                ) { showNote = !showNote }
            }

            AnimatedVisibility(showNote) {
                BasicTextField(
                    value = note,
                    onValueChange = { note = it.take(500) },
                    modifier = Modifier.fillMaxWidth().padding(top = 10.dp)
                        .clip(RoundedCornerShape(16.dp))
                        .border(1.dp, Outline, RoundedCornerShape(16.dp))
                        .padding(horizontal = 14.dp, vertical = 12.dp),
                    textStyle = MaterialTheme.typography.bodyMedium.copy(color = Paper),
                    maxLines = 3,
                    cursorBrush = SolidColor(Amber),
                    decorationBox = { field ->
                        if (note.isEmpty()) {
                            Text("Anything worth remembering", color = Slate, style = MaterialTheme.typography.bodyMedium)
                        }
                        field()
                    },
                )
            }

            Keypad(
                Modifier.padding(top = 14.dp),
                onKey = { key -> amount = applyKey(amount, key) },
            )

            PrimaryButton(
                text = if (expense == null) "Print it to the tape" else "Save the line",
                modifier = Modifier.fillMaxWidth().padding(top = 14.dp),
                busy = busy,
            ) {
                val rate = if (paidIn == currency) null else fx.rateBetween(paidIn, currency, snapshot)
                error = when {
                    parsed == null || parsed.signum() <= 0 -> "Enter an amount above zero."
                    title.isBlank() -> "What was it for?"
                    paidIn != currency && rate == null -> "No rate is available for $paidIn to $currency."
                    else -> null
                }
                if (error == null) scope.launch {
                    busy = true
                    val draft = ExpenseDraft(
                        title.trim(),
                        converted ?: parsed!!.setScale(2, RoundingMode.HALF_UP),
                        if (paidIn == currency) null else parsed!!.setScale(2, RoundingMode.HALF_UP),
                        paidIn.takeIf { it != currency }, rate, category, spentOn,
                        note.trim().ifBlank { null },
                    )
                    onSave(draft).onSuccess {
                        onRememberCurrency(paidIn)
                        onDismiss()
                    }.onFailure { error = it.message ?: "Could not save that line." }
                    busy = false
                }
            }

            if (expense != null && onDelete != null) {
                TextButton(
                    onClick = { confirmDelete = true },
                    modifier = Modifier.fillMaxWidth().padding(top = 2.dp),
                ) { Text("Delete this line", color = ErrorRed, fontWeight = FontWeight.Bold) }
            }
            Spacer(Modifier.height(16.dp))
        }
    }

    if (confirmDelete && onDelete != null) AlertDialog(
        onDismissRequest = { confirmDelete = false },
        shape = RoundedCornerShape(28.dp),
        containerColor = CardNavy,
        title = { Text("Delete this line?", style = MaterialTheme.typography.headlineSmall) },
        text = { Text("This cannot be undone.", color = Slate) },
        confirmButton = {
            TextButton(onClick = {
                scope.launch {
                    onDelete().onSuccess { onDismiss() }.onFailure { error = it.message }
                    confirmDelete = false
                }
            }) { Text("Delete", color = ErrorRed, fontWeight = FontWeight.Bold) }
        },
        dismissButton = { TextButton(onClick = { confirmDelete = false }) { Text("Keep", color = Steel) } },
    )
    if (discard) AlertDialog(
        onDismissRequest = { discard = false },
        shape = RoundedCornerShape(28.dp),
        containerColor = CardNavy,
        title = { Text("Discard this line?", style = MaterialTheme.typography.headlineSmall) },
        text = { Text("It has not been printed to the tape.", color = Slate) },
        confirmButton = { TextButton(onClick = onDismiss) { Text("Discard", color = ErrorRed, fontWeight = FontWeight.Bold) } },
        dismissButton = { TextButton(onClick = { discard = false }) { Text("Keep editing", color = Steel) } },
    )
}

/**
 * Applies one keypad press to the amount. Kept pure and separate so the rules —
 * one decimal point, at most two decimals, no runaway leading zeroes — are
 * readable in one place rather than scattered through the key handlers.
 */
internal fun applyKey(current: String, key: String): String = when {
    key == "⌫" -> current.dropLast(1)
    key == "." -> if (current.contains('.')) current else if (current.isEmpty()) "0." else "$current."
    current.substringAfter('.', "").length >= 2 && current.contains('.') -> current
    current == "0" -> key
    current.length >= 12 -> current
    else -> current + key
}

private val KEYS = listOf("1", "2", "3", "4", "5", "6", "7", "8", "9", ".", "0", "⌫")

@Composable
private fun Keypad(modifier: Modifier = Modifier, onKey: (String) -> Unit) {
    Column(modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(10.dp)) {
        KEYS.chunked(3).forEach { row ->
            Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                row.forEach { key -> Key(key, Modifier.weight(1f)) { onKey(key) } }
            }
        }
    }
}

@Composable
private fun Key(label: String, modifier: Modifier = Modifier, onClick: () -> Unit) {
    val interactionSource = remember { MutableInteractionSource() }
    val pressed by interactionSource.collectIsPressedAsState()
    Box(
        modifier.height(54.dp)
            .clip(RoundedCornerShape(18.dp))
            .background(if (pressed) Amber.copy(alpha = .22f) else CardRaised)
            .clickable(interactionSource = interactionSource, indication = null, onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            label,
            color = if (pressed) Amber else Paper,
            fontFamily = PlexMono,
            fontSize = 22.sp,
        )
    }
}

@Composable
private fun DatePill(value: LocalDate, modifier: Modifier = Modifier, onChange: (LocalDate) -> Unit) {
    var open by remember { mutableStateOf(false) }
    val today = LocalDate.now()
    FilterPill(
        when (value) {
            today -> "Today"
            today.minusDays(1) -> "Yesterday"
            else -> com.wayfare.app.core.shortDate(value)
        },
        value != today,
        modifier,
    ) { open = true }
    // A line always happened on some day, so clearing the date is not offered.
    if (open) DatePickerSheet(value, { open = false }, allowClear = false) { picked ->
        picked?.let(onChange)
        open = false
    }
}

@Composable
private fun CurrencyPill(value: String, modifier: Modifier = Modifier, onChange: (String) -> Unit) {
    var open by remember { mutableStateOf(false) }
    Box(modifier) {
        FilterPill(value, false, Modifier.fillMaxWidth()) { open = true }
        DropdownMenu(expanded = open, onDismissRequest = { open = false }) {
            FxRepository.CURRENCIES.forEach { currency ->
                DropdownMenuItem(text = { Text(currency) }, onClick = { onChange(currency); open = false })
            }
        }
    }
}

/** A line someone else added: readable, never editable. */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ReadOnlyExpenseSheet(expense: Expense, currency: String, payer: String, onDismiss: () -> Unit) {
    FormSheet(
        onDismissRequest = onDismiss,
        title = { Text(expense.title) },
        text = {
            Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
                FigureWithQualifier(money(expense.amount, currency), null, size = 38.sp)
                Text(
                    "Printed by $payer. Only the traveller who added a line can change it.",
                    color = Slate,
                    style = MaterialTheme.typography.bodyMedium,
                )
                MonoLabel(
                    "${expense.category.label.uppercase()} · ${com.wayfare.app.core.tapeDayLabel(expense.spentOn)}",
                    color = Slate,
                )
                expense.originalCurrency?.let {
                    Text("Originally ${expense.originalAmount} $it", color = Slate, style = MaterialTheme.typography.bodyMedium)
                }
                expense.note?.let { Text(it, color = Paper, style = MaterialTheme.typography.bodyMedium) }
            }
        },
        confirmButton = { TextButton(onClick = onDismiss) { Text("Close", color = Amber, fontWeight = FontWeight.Bold) } },
    )
}

/** A line that never reached the server: kept, explained, retried or discarded. */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun UnsyncedExpenseDialog(
    expense: Expense,
    currency: String,
    onDismiss: () -> Unit,
    onRetry: () -> Unit,
    onDiscard: () -> Unit,
) {
    val failed = expense.syncState == com.wayfare.app.core.SyncState.Failed
    FormSheet(
        onDismissRequest = onDismiss,
        title = { Text(if (failed) "This line did not print" else "Still printing") },
        text = {
            Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
                DottedLeaderRow(expense.title, money(expense.amount, currency))
                Text(
                    if (failed) {
                        "It is kept here so nothing is lost, but it is not counted in the trip total " +
                            "and other travellers cannot see it yet."
                    } else {
                        "It is saved on this phone and will sync as soon as there is a connection. " +
                            "You can edit it once it lands."
                    },
                    color = Slate,
                    style = MaterialTheme.typography.bodyMedium,
                )
                expense.syncError?.let { Notice(it, error = failed) }
            }
        },
        confirmButton = { PrimaryButton("Try again", onClick = onRetry) },
        dismissButton = {
            Row {
                if (failed) TextButton(onClick = onDiscard) { Text("Discard", color = ErrorRed, fontWeight = FontWeight.Bold) }
                TextButton(onClick = onDismiss) { Text("Close", color = Steel) }
            }
        },
    )
}
